src/peer_connection.rs

```diff
@@ -27,40 +28,96 @@ impl Handshake {
 }
 
 pub struct PeerConnection {
-    stream: TcpStream,
+    // stream: TcpStream,
     bitfield: Vec<u8>,
+    outgoing_tx: mpsc::Sender<message::Message>,
 }
 
 pub struct PeerConnections {
     connections: Vec<PeerConnection>,
+    peer_incoming_rxs: Vec<mpsc::Receiver<message::Message>>,
 }
 
 impl PeerConnection {
-    pub fn new(stream: TcpStream) -> PeerConnection {
-        PeerConnection {
-            stream,
+    pub fn new(peer: &tracker::Peer,
+               metadata: &parser::TorrentMetadata) -> (PeerConnection, mpsc::Receiver<message::Message>) {
+        let (incoming_tx, incoming_rx) = mpsc::channel(1024);
+        let (outgoing_tx, outgoing_rx) = mpsc::channel(1024);
+
+        let socket_addr = SocketAddr::from(SocketAddrV4::new(peer.ip, peer.port));
+        tokio::spawn(peer_connection_background(socket_addr.clone(), metadata.clone(), incoming_tx, outgoing_rx));
+
+        (PeerConnection {
+            // stream,
             bitfield: Vec::new(),
+            outgoing_tx,
+        }, incoming_rx)
+    }
+
+}
+
+async fn peer_connection_background(socket_addr: SocketAddr,
+                                    metadata: parser::TorrentMetadata,
+                                    mut incoming_tx: mpsc::Sender<message::Message>,
+                                    mut outgoing_rx: mpsc::Receiver<message::Message>)
+    -> Result<(), String> {
+    let mut tcp_stream = perform_handshake(&socket_addr, &metadata).await.map_err(|e| e.to_string())?;
+
+    loop {
+        while let Some(outgoing_message) = outgoing_rx.try_recv().ok() {
+            println!("[unimplemented] send_message {:?}", outgoing_message);
+        }
+
+        let message = receive_message(&mut tcp_stream).await?;
+        println!("[info] received peer message: {:?}", message);
+
+        incoming_tx.send(message).await.unwrap();
```

Comment:
Similarly, we don't care what the message looks like here, we just send whatever the incoming message is and then in `download.rs` the incoming_rx will receive the message and parse/process it there?